[DO NOT MERGE] Initial support for backup of D2D-only apps

t-m-w opened the pull request:
(It was proposed that a more minimal initial-support PR be opened, vs #473, to make it easier to test functionality on-device.)

Allow backup of apps that would otherwise only support device-to-device migration. This is an initial-support patch to help determine the viability of this approach.

Known issues / TODO:
* System-scheduled backups will not handle D2D-only apps, unless accompanied by a framework change forcing OperationType.MIGRATION. Backups triggered by the connection of a USB device or by Seedvault's StorageBackupService (files) scheduling are not affected, so they *will* back up D2D-only apps as expected; otherwise, the user may need to perform a backup manually via Backup Now.
* Apps with `allowBackup="false"` will appear in Backup Status under "Installed Apps" rather than "Apps that do not allow data backup", and their status will always be blank until they have been backed up. If they are not eligible for migration, it will never change.

Other notes:
* The unit test for excluding the Storage Plugin provider from backups was discussed, deemed unnecessary, and removed.

Co-authored-by: Oliver Scott <olivercscott@gmail.com>
Change-Id: I5a23d68be66f7d8ed755f2bccb9570ab7be49356

grote committed:
Pretend to be a device to device data transport

Change-Id: Ieab13cb15b64e09798ec0c1081d3b3cd975e0811

t-m-w committed:
Initial support for backup of D2D-only apps

Allow backup of apps that would otherwise only support device-to-device
migration. This is an initial-support patch to help determine the
viability of this approach.

Known issues / TODO:
* System-scheduled backups will not handle D2D-only apps, unless
  accompanied by a framework change forcing OperationType.MIGRATION.
  Backups triggered by the connection of a USB device or by Seedvault's
  StorageBackupService (files) scheduling are not affected, so they
  *will* back up D2D-only apps as expected; otherwise, the user may
  need to perform a backup manually via Backup Now.
* Apps with `allowBackup="false"` will appear in Backup Status under
  "Installed Apps" rather than "Apps that do not allow data backup",
  and their status will always be blank until they have been backed
  up. If they are not eligible for migration, it will never change.

Other notes:
* The unit test for excluding the Storage Plugin provider from backups
  was discussed, deemed unnecessary, and removed.

Co-authored-by: Oliver Scott <olivercscott@gmail.com>
Change-Id: I5a23d68be66f7d8ed755f2bccb9570ab7be49356

chirayudesai commented:
My goal was to have this be enough to actually test this functionally on the device, and we could continue further refinements on the original PR

chirayudesai commented:
14597: [D2D] System-scheduled backups act as migrations | https://review.calyxos.org/c/CalyxOS/platform_frameworks_base/+/14597

t-m-w commented:
Adding my testing notes...

Apps for testing:
- Tested using three app backup eligibility categories: allowed by app, disallowed by app but migration-eligible, disallowed by app and not migration-eligible. For each category, I chose two apps, and excluded one from backup. For testing, I selected the following apps:
  - [OpenTasks](https://f-droid.org/en/packages/org.dmfs.tasks)
  - _[Shopping List (SECUSO)](https://f-droid.org/en/packages/privacyfriendlyshoppinglist.secuso.org.privacyfriendlyshoppinglist)_
  - [Peanut Encryption](https://f-droid.org/en/packages/peanutencryption.peanutencryption)
  - _[Tor Browser](https://www.torproject.org/download/#android)_
  - [Safe Notes - Encrypted Notepad](https://f-droid.org/en/packages/com.trisven.safenotes)
  - _[Bromite](https://github.com/bromite/bromite)_
- Exclude from backup, manually: Shopping List (SECUSO), Tor Browser, Bromite
- Category: allowBackup="true"
  - OpenTasks
    - Task: `Test`
  - (ADD USER EXCLUDE!) Shopping List by SECUSO
    - List: `Test`
      - Item: `Test`
- Category: allowBackup="false"
  - Sub-category: SDK <=30: system will exclude from backup
    - [Peanut Encryption (peanutencryption.peanutencryption) 1.3](https://f-droid.org/en/packages/peanutencryption.peanutencryption/)
      - Password: `test123!`
        - Entry: `test`, `test`
    - (ADD USER EXCLUDE!) Tor Browser 99.0.0b3
      - Settings > Configure a bridge > meek-azure
  - Sub-category: SDK >=31: system will include in backup anyway
    - [Safe Notes - Encrypted Notepad (com.trisven.safenotes) 2.1.0+](https://github.com/keshav-space/safenotes/releases)
      - Password: `test123!`
    - (ADD USER EXCLUDE!) Bromite
      - Bookmark: `about:blank`

Verify:
- Data gets backed up for OpenTasks (_allowed by app_), Safe Notes (_disallowed by app, but migration allows_)
- Data does not get backed up/restored for Shopping List by SECUSO (_excluded, allowed by app_), Tor Browser (_user-excluded, disallowed by app, and not migration-eligible_), Bromite (_user-excluded, disallowed by app, but migration-eligible_), Peanut Encryption (_disallowed by app, not migration-eligible_).

Ways of triggering a backup:
- Scheduled backup via `adb shell bmgr run` (K/V only?) / `adb shell bmgr backupnow --all` (all).
- Manual backup via Backup Now.
- Backup via plugging in USB.

Other things to look at:
- Force Stop behavior
  - Does Force Stop of an app influence backup behavior? Does it still get backed up? It shouldn't, apparently.
- UI: Backup status screen: Before test.
- UI: Backup status screen: During test.
- UI: Backup status screen: After test.
- UI: Backup errors reported.
- UI: Restoration errors reported.

chirayudesai committed:
Bump version to 13-3.2

* Spread out the version code a bit to allow for more
  releases per one name

Change-Id: I70b6ba92e359a1f229ee470f4032c04b9fd25e5b

chirayudesai committed:
[DO NOT MERGE] Bump version and set property for testing

* Bump version
  * 33030020 -> 13-3.2
  * 33030021 -> This, d2d test #478
  * 33030022 -> Next, to go back to normal build after testing #489
* Set the testing property to true so that the system
  sends us the right app list
  14597: Optionally have System-scheduled backups act as migrations | https://review.calyxos.org/c/CalyxOS/platform_frameworks_base/+/14597

Change-Id: I278091b6659db095716e01b6c3894ce345219283

nettnikl commented:
Hi @chirayudesai @t-m-w , could you update on the current status? What is the current status? Needs testing, needs review, needs to have some known bugs resolved, ...?

chirayudesai commented:
> Hi @chirayudesai @t-m-w , could you update on the current status? What is the current status? Needs testing, needs review, needs to have some known bugs resolved, ...?

Hey. All 3 of those are currently on our list. For testing we have a plan to release it in a way that would make public testing easier, but that'll take some time to setup.

stevesoltys commented:
Resolved via #562.
